docs/designs/data_transform.md

```diff
@@ -1,16 +1,38 @@
+* The transform logic is incomplete at the time when user develops the pipeline. The statistical value on the entire dataset at runtime is necessary to make the transform logic concrete.  
+
+## Related Work
+
+### TensorFlow Transform
+
+[TensorFlow Transform](https://www.tensorflow.org/tfx/transform/get_started) is the open source solution for data transform in [TensorFlow Extended](https://www.tensorflow.org/tfx/guide). Users need write a python function 'preprocess_fn' to define the preprocess logic. The preprocessing function contains two group of API calls: TensorFlow Transform Analyzers and TensorFlow Ops. Analyzer will do the statistical work on the training dataset once and convert to constant tensor. And then the statistical value and TensorFlow Ops will make the concrete transform logic as the TensorFlow Graph to convert the record one by one. The graph will be used both for training and serving.  
+From users perspective, SQLFlow users prefer to write SQL instead of python. It's user unfriendly to SQLFlow users if we integrate TF Transform with SQLFlow directly.  
+
+### SQL
+
+User can write SQL to do the analysis and transform work with [built-in functions](https://www.alibabacloud.com/help/doc-detail/96342.htm?spm=a2c63.p38356.b99.111.27e27309rgC5m1) or [UDF(User Defined Function)](https://www.alibabacloud.com/blog/udf-development-guide-with-maxcompute-studio_594738). But we can't use SQL to transform the data for serving. We need implement the transform code again using C++ or other high performance language. It may bright the training/serving skew.  
+
+### Internal System
+
+The library in the internal system is configure driven. It contains some primitive transform ops and user compose the transform logic with configuration file. The inference engine is compiled with the internal library. User need both feature engineering configuration files and SavedModel to deploy the model for serving.  
+
+## Our Approach
+
+From the point of view from SQLFlow, SQL can naturally support statistical work just like the analyzer. [Feature column API](https://tensorflow.google.cn/api_docs/python/tf/feature_column) and [keras preprocessing layer](https://github.com/tensorflow/community/pull/188) can take charge of transform logic as transformer. For dense column, we can use numeric_column and pass a user defined function *normalizer_fn* to convert the column value. For sparse column, we can use embedding_column to map the sparse value to embedding vector or use cross_column to do the feature crossing. We plan to use SQL and feature column/keras preprocessing layer together to do the data transform work.  
```

Comment:
I was wondering whether there is sparse column in the table? Do the sparse column represent the column with ids string like ["1,2,4", "1,5,6,7"]. But it is also a dense column as string format. What's more, tf-serving does not support varied length tensor.

Reply:
The sparse column represent the column feed into categorical_column_with_xxx or sequence_categorical_column_with_xxx.